Circuit and Cell Events (Draft)

teor2345 opened the pull request:
Event call sites for your comments: this branch will be updated with the event content tomorrow.

teor2345 committed:
Add new CELL and CIRCUIT events, and placeholders for other expected events

Part of #368.

teor2345 committed:
Add a placeholder function for the new CIRCUIT event

Refactor the old CIRCUIT event so it's generated via the new event's
function.

Part of #368.

teor2345 committed:
Add a placeholder function for the new CELL event

Add some missing entry points, we'll filter them out in the PrivCount
code if we need to.

Part of #368.

teor2345 committed:
Remove all instances of relay_command

It's now cell->command.

Part of #368.

robgjansen commented on src/or/relay.c:
```diff
@@ -2784,8 +2821,23 @@ append_cell_to_circuit_queue(circuit_t *circ, channel_t *chan,
+  if (circ->marked_for_close) {
+
+    /* This was missing from the original cell code. Why? */
+    /* This cell is never actually sent: ignore it in the PrivCount code */
```
I think if we didn't actually send a cell, then we should not report that is was sent. That might mean not emitting an event here, or emitting it with a special flag to indicate that it wasn't sent.

teor2345 replied:
Coded and documented in ae25634dfb.

robgjansen commented on src/or/relay.c:
```diff
@@ -221,14 +221,51 @@ circuit_receive_relay_cell(cell_t *cell, circuit_t *circ,
+  if (circ->marked_for_close) {
+
+    /* This was missing from the original cell code. Why? */
+    /* This cell is never actually processed: ignore it in the PrivCount code?
```
We didn't process this one because the circuit was marked, but we *did* receive it; so we should probably emit the event (possibly with a flag to note that it wasn't processed and why).

teor2345 replied:
Coded and documented in ae25634dfb.

robgjansen commented on src/or/relay.c:
```diff
@@ -221,14 +221,51 @@ circuit_receive_relay_cell(cell_t *cell, circuit_t *circ,
     log_warn(LD_BUG,"relay crypt failed. Dropping connection.");
+
+    /* This was missing from the original cell code. Why? */
+    /* This cell is never actually processed: ignore it in the PrivCount code?
```
We *did* actually receive the cell, we just couldn't decrypt it. I would say we should emit the event, possibly with the reason why it wasn't processed.

teor2345 replied:
Coded and documented in f6ce047bde

robgjansen commented on src/or/relay.c:
```diff
@@ -221,14 +221,51 @@ circuit_receive_relay_cell(cell_t *cell, circuit_t *circ,
+      control_event_privcount_circuit_cell(circ->n_chan, circ, cell,
+                                           PRIVCOUNT_CELL_RECEIVED);
+    } else {
+      /* This was missing from the original cell code. Why? */
```
We were only concerned about non-origin cells; but yes, we should probably emit the cell with the correct flags indicating they are origin cells and do any filtering in PrivCount.

teor2345 replied:
Documented in a6ba447d31

robgjansen commented on src/or/command.c:
```diff
@@ -255,6 +255,13 @@ command_process_create_cell(cell_t *cell, channel_t *chan)
            " ignoring.", (int)cell->command,
            channel_get_actual_remote_descr(chan));
+
+    /* PrivCount will just ignore this, but do it anyway for consistency */
```
For these first cases where this function returns, we should emit the cell event but probably flag that it is an error case. I guess that will be captured by the fact that there is not yet a circuit associated with the cell.

Note that we *can't* move `control_event_privcount_circuit_cell` earlier in this function before any code would cause a return, and then only call it once - because the last call to it on line 357 has a new circuit associated with it and a circuit id (which we want to log) and the rest don't have a circuit yet.

teor2345 replied:
Coded and documented in b88c6c54fc

robgjansen commented on src/or/command.c:
```diff
@@ -534,6 +607,13 @@ command_process_relay_cell(cell_t *cell, channel_t *chan)
            "(%s) failed. Closing.",
            direction==CELL_DIRECTION_OUT?"forward":"backward");
+
+    /* This was missing from the original cell code. Why? */
```
Because `circuit_receive_relay_cell` was just called above, and that function would handle any additional places where the cell event should get emitted. We don't want to emit multiple cell events for the same cell.

teor2345 replied:
Coded and documented in e11c927dd7

robgjansen commented on src/or/command.c:
```diff
@@ -473,11 +521,24 @@ command_process_relay_cell(cell_t *cell, channel_t *chan)
               (unsigned)cell->circ_id,
               channel_get_canonical_remote_descr(chan));
+
+    /* PrivCount will just ignore this, but do it anyway for consistency */
```
Note that `control_event_privcount_circuit_cell ` should only be called in this function in places where this function returns before calling `circuit_receive_relay_cell`; if `circuit_receive_relay_cell` is called, the cell event will be emitted there instead.

teor2345 replied:
Coded and documented in e11c927dd7

robgjansen reviewed (commented):
Looks good overall. There is one call site that was added that would cause double counting relay-type cells (line 613 of `command.c`), but I think the other additions are OK.

I think that most of the call site omissions were because I was ignoring error cases and origin cells, but it's a good idea to keep those in and let PrivCount filter as desired.

teor2345 committed:
Always send cell events when the circuit is NULL

There won't be any circuit fields in these events, so ounters can filter
them out.

Part of #368.

teor2345 committed:
Mark circuits for close before sending the cell event

And explain how that is handled in the event function

teor2345 committed:
When a received relay cell fails decryption, add fields to the event

teor2345 committed:
Document that cell events include origin cells

teor2345 committed:
Don't send duplicate cell events in command_process_relay_cell()

teor2345 committed:
Consistently order circuit actions, channel actions, and cell events

Always mark circuts for close before sending a cell event.

Always send destroy cells on channels after sending the received cell's event

teor2345 commented:
See also 	7bddab1, which makes sure that:
* we always send the cell event *after* marking the circuit for close, so it can be filtered out, and
* we always send the cell event for the channel destroy cell *after* the received cell that triggered the channel destroy

teor2345 committed:
Add extra circuit end events for origin circuits

Continue to send legacy circuit events just as they were sent in previous
versions.

teor2345 committed:
Add validation for control_event_privcount_circuit_cell() arguments

teor2345 committed:
Add fields to the cell event

teor2345 commented:
I made sure that the legacy circuit event is sent for exactly the same circuits in 84b5b3d.

In d12c268 and 511ef91, I added the fields to the cell event.
I tried to keep the order you have in the existing code.
The only field I had trouble with was the relay command.
See the code for details.

I'll do the circuit event fields on Monday.
And then I will write the parsing code, and test both the events.

teor2345 committed:
Consistently use inbound/outbound and cells/bytes in identifiers

teor2345 committed:
Fix line lengths after the search and replace

teor2345 committed:
Initialise try_relay_command

This is safer if more cases are added later.

teor2345 committed:
Add legacy circuit fields to the new circuit close event

Also refactor out common cell and circuit fields into their own
function.

teor2345 committed:
Rename buf to fingerprint in privcount_list_hs_v2_intro_point_fingerprints()

teor2345 committed:
fixup! Add legacy circuit fields to the new circuit close event

teor2345 committed:
Add the fingerprint and relay flags for previous and next nodes

Don't add nickname and IPv4 address, they are redundant (and can be
retrieved from the consensus if needed).

Don't add anything for this relay: the aggregator already knows this
relay's flags and fingerprint.
(And a lot of other info, check the get_* functions for details.)

teor2345 committed:
fixup! Add the fingerprint and relay flags for previous and next nodes

teor2345 committed:
Add some missing consts to the HSDir Store event function

teor2345 committed:
Put the legacy exit bandwidth fields at the end of the event

teor2345 committed:
Add state, purpose, and spliced rend circuit IDs to the circuit event

Don't put channel IDs in each cell event.
Rearrange some fields for clarity.

Part of #368.

teor2345 commented:
Hi @robgjansen,
I have added all the fields I think we will need to the circuit event:
* 37215f9  adds the fields from the legacy event
* 9240626 adds fingerprints and relay flags for the previous and next nodes
  (I didn't bother to add the nickname and IP addresses, they seemed redundant)
* 4f5756d adds state, purpose, and spliced rend connection IDs

As discussed, this is enough information to count single onion services.

I'm finished adding fields to these events, I will write the PrivCount parsing code and do some testing tomorrow.

teor2345 committed:
Make NULL handling better, again

teor2345 committed:
Add cell counts and directory byte counts to the circuit event

Also refactor the calling code so it's much simpler, and updates the
counters for legacy and new events.

teor2345 committed:
fixup! Put the legacy exit bandwidth fields at the end of the event

teor2345 committed:
Cleanup some whitespace

teor2345 committed:
Remove some unused functions

teor2345 commented:
It turns out I wasn't quite done:
* b0f3bd6 adds cell and directory byte counts to the circuit event
  (that's all the counts we'll ever have, we need to use cell counts or connections for anything else, see privcount/privcount#369)
* there were also a few bugs and untidy bits of code

I think it's actually ready now, so I applied all my fixups to their base commits.
See https://github.com/teor2345/tor-privcount/tree/circuit-cell-squashed for new work.

teor2345 commented:
Opened https://github.com/privcount/tor/pull/10 so we merge this.
